rfcs/0003-build-contents-diff.md

```diff
@@ -0,0 +1,160 @@
+Compares the two specified builds
+ 
+Three use cases that this feature should ideally satisfy:
+ 
+- As a DevOps engineer, when I see a new build has been created in a lower environment, I want to evaluate what benefit it provides to my team for me to promote this build to an upper environment. 
+ 
+- As a DevOps engineer or App Developer, when I am told there is a problem with a specific build (new build failing in CI or old build failing in production), I want to know what’s different about the problem build than the last one that worked. 
+
+- As an App Developer, I want to see the difference between the commit I built locally with pack yesterday and the build that kpack built for me today based on my commit. 
+ 
+Implications: 
+- User will need to compare their most recent build to the older build in production
+i.e. builds being compared are non-adjacent 
+- Because the builds are non-adjacent, many things are likely to have changed in the contents and there will have been many “Reasons” that caused the intervening builds. Therefore, it doesn’t make a lot of sense to print the “Reason” as part of the diff
+- I assume that it is easy / possible for users to access the build number associated with a container in production (perhaps there is another build identifier we could use?)
+
+
+
+## Prior Art 
+Differences from the reason diff RFC that’s been added to `kp build status`:
+COMMIT is the same 
+CONFIG is the same
+BUILDPACK is different because it prints the Bill of Materials which represents the results of the buildpacks used instead of only the buildpack/s that caused the new build to kick off
+STACK is the same now, but will be different once we have more stack metadata such as the dpackage list (and someday a CVE list) that we can print
+ 
+Competitor examples: 
+https://github.com/GoogleContainerTools/container-diff
+
+## Complexity/Risks
+- Would this be helpful for diagnosing failed builds or builds that don't start? If so, how can we surface this info without a BoM being generated? 
+- How does all this show up in CI, if at all? 
+
+## Alternatives
+- Add information about build contents diff to the kp build status command
+  - Pros:
+Reduce the CLI surface area 
+May be a more intuitive location
+  - Cons:
+Makes a command with long output even longer 
+- Print / pipe out two YAML or JSON files that represent the build contents and allow the user to diff them with the tool of their choice 
+  - Pros:
+Export two image digests and compare them 
+  - Cons: 
+More work for the user
+
+- Everything the same as the proposal, except show the things that didn’t change only with a verbose flag 
+ 
+Deliver a feature like this as part of a separate tool:
+- Registry  
+- `Pack`
+
+
+## Mockups 
+Notes on WIP mockup
+- don’t really love the all caps headers (maybe bold or other color?)
+- should there be some kind of header info to set the stage for the diff? 
+ 
+------------------
+``` 
+$kp build diff petclinic 37 13
```

Comment:
As this information is largely the result of a built image, it seems like we should disconnect it from the "Build Reason" Headers. This would enable us to instead focus on displaying the unique information in the the bill of materials.

Reply:
This would also enable us to `diff` against non kpack built images theoretically.

Reply:
ok i hear you, let me try that out